File: 630_schedule_course/schedule_course.cpp
/*
 * @Date: 2021-12-14 05:54:56
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-12-14 06:07:19
 */

#include <cassert>
#include <queue>
#include <vector>

using namespace std;

class Solution {
 public:
  int scheduleCourse(vector<vector<int>> courses) {
    sort(courses.begin(), courses.end(),
         [](const auto& c0, const auto& c1) { return c0[1] < c1[1]; });

    priority_queue<int> q;
    // 优先队列中所有课程的总时间
    int total = 0;

    for (const auto& course : courses) {
      int ti = course[0], di = course[1];
      if (total + ti <= di) {
        total += ti;
        q.push(ti);
      } else if (!q.empty() && q.top() > ti) {
        total -= q.top() - ti;
        q.pop();
        q.push(ti);
      }
    }

    return q.size();
  }
};

int main() {
  assert(Solution().scheduleCourse(vector<vector<int>>{
             {100, 200}, {200, 1300}, {1000, 1250}, {2000, 3200}}) == 3);
  assert(Solution().scheduleCourse(vector<vector<int>>{{1, 2}}) == 1);
  assert(Solution().scheduleCourse(vector<vector<int>>{{3, 2},{4,3}}) == 0);
}

File: rust/src/630_schedule_course/schedule_course.rs
/*
 * @Date: 2021-12-14 05:54:56
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-12-14 06:10:54
 */

pub fn schedule_course(courses: Vec<Vec<i32>>) -> i32 {
    use std::collections::BinaryHeap;
    let mut c = courses;
    c.sort_by(|a, b| i32::cmp(&a[1], &b[1]));

    let mut heap = BinaryHeap::new();
    let mut total = 0;

    for cource in c {
        let duration = cource[0];
        let last_day = cource[1];
        if total + duration <= last_day {
            total += duration;
            heap.push(duration);
        } else if !heap.is_empty() && heap.peek().unwrap() > &duration {
            total -= heap.pop().unwrap() - duration;
            heap.push(duration);
        }
    }
    heap.len() as i32
}

fn main() {
    assert_eq!(
        schedule_course(vec![
            vec![100, 200],
            vec![200, 1300],
            vec![1000, 1250],
            vec![2000, 3200]
        ]),
        3
    );
    assert_eq!(schedule_course(vec![vec![1, 2]]), 1);
    assert_eq!(schedule_course(vec![vec![3, 2], vec![4, 3]]), 0);
}
